clients/venice-admin-tool/src/main/java/com/linkedin/venice/KafkaTopicDumper.java

```diff
@@ -234,66 +226,200 @@ public KafkaTopicDumper(
         }
       }
     }
+    if (!(logMetadata || logDataRecord)) {
+      setupDumpFile();
+    }
+  }
 
-    PubSubTopicPartition partition =
-        new PubSubTopicPartitionImpl(pubSubTopicRepository.getTopic(topicName), partitionNumber);
+  // Constructor for testing purpose only
+  KafkaTopicDumper(PubSubConsumerAdapter consumer, PubSubTopicPartition topicPartition, int maxConsumeAttempts) {
+    this.topicPartition = topicPartition;
+    this.consumer = consumer;
+    this.maxConsumeAttempts = maxConsumeAttempts;
+    this.keySchemaStr = null;
+    this.keySchema = null;
+    this.keyReader = null;
+    this.latestValueSchemaStr = null;
+    this.allValueSchemas = null;
+    this.parentDirectory = null;
+    this.logMetadata = false;
+    this.logDataRecord = false;
+    this.logRmdRecord = false;
+    this.logTsRecord = false;
+    this.isChunkingEnabled = false;
+    this.compressorFactory = new CompressorFactory();
+    this.compressor = compressorFactory.getCompressor(CompressionStrategy.NO_OP);
+    this.chunkKeyValueTransformer = null;
+    this.chunkedKeySuffixDeserializer = null;
+    this.manifestSerializer = null;
+  }
 
+  /**
+   * Calculates the starting offset for consuming messages from a PubSub topic partition.
+   *
+   * <p>This method determines the appropriate starting offset based on the provided starting
+   * offset or timestamp. If a specific {@code startingTimestamp} is provided, it attempts to find
+   * the offset corresponding to that timestamp. If no offset is found for the timestamp, it throws
+   * a {@link PubSubClientException}. The calculated starting offset will always be greater than
+   * or equal to the partition's beginning offset to ensure it is within the valid range.
+   *
+   * @param consumer the {@link PubSubConsumerAdapter} used to fetch offsets
+   * @param partition the {@link PubSubTopicPartition} identifying the topic and partition
+   * @param startingOffset the default starting offset to use if no timestamp is provided
+   * @param startingTimestamp the starting timestamp (epoch in milliseconds) to locate an offset.
+   *                          If set to -1, the method uses the provided {@code startingOffset}.
+   * @return the calculated starting offset for the partition
+   * @throws PubSubClientException if no offset is found for the provided timestamp
+   */
+  static long calculateStartingOffset(
+      PubSubConsumerAdapter consumer,
+      PubSubTopicPartition partition,
+      long startingOffset,
+      long startingTimestamp) {
+    if (startingTimestamp != -1) {
```

Comment:
I intentionally kept this behavior because sometimes the startDate can overshoot the endOffset, and in such cases, the fallback would be the provided startingOffset. I personally prefer this approach, but if you think it might cause confusion, I can update it.

Reply:
Yeah, it's confusing

Reply:
Done